### ISSUE TITLE
0007392:
SQL Foreign Key Error on save of Contact

### ISSUE THREAD

**Issue by Gloirin** [tine20/Tine-2.0-Open-Source-Groupware-and-CRM#3740]
**Reported by TheSchmuck on 9 Nov 2012 04:19**

**Version:** Joey (2012.10.1)

I began using Tine in around July 2010 and have not upgraded it in the time since. Recently I have downloaded the latest release (Joey) and used the setup.php to upgrade the applications.

Now whenever trying to save a new contact, or a new task I receive an SQL error.

During the upgrade process, when attempting to upgrade all applications together (which is the default), I receive the following error:

SQLSTATE[HY000]: General error: 1005 Can&#039;t create table &#039;tine20_2010.tine20_container_content&#039; (errno: 150)

.../library/Zend/Db/Statement.php(284): Zend_Db_Statement_Pdo-&gt;_execute()
.../library/Zend/Db/Adapter/Abstract.php(468): Zend_Db_Statement-&gt;execute()
.../library/Zend/Db/Adapter/Pdo/Abstract.php(238): Zend_Db_Adapter_Abstract-&gt;query()
.../Setup/Backend/Abstract.php(207): Zend_Db_Adapter_Pdo_Abstract-&gt;query()
.../Setup/Backend/Abstract.php(251): Setup_Backend_Abstract-&gt;execQueryVoid()
.../Tinebase/Setup/Update/Release5.php(410): Setup_Backend_Abstract-&gt;createTable()
.../Setup/Controller.php(435): Tinebase_Setup_Update_Release5-&gt;update_9()
.../Setup/Controller.php(326): Setup_Controller-&gt;updateApplication()
.../Setup/Frontend/Json.php(118): Setup_Controller-&gt;updateApplications()
[internal function]: Setup_Frontend_Json-&gt;updateApplications()
.../library/Zend/Server/Abstract.php(232): call_user_func_array()
.../Zend/Json/Server.php(558): Zend_Server_Abstract-&gt;_dispatch()
.../Zend/Json/Server.php(197): Zend_Json_Server-&gt;_handle()
.../Setup/Server/Json.php(77): Zend_Json_Server-&gt;handle()
.../Setup/Core.php(89): Setup_Server_Json-&gt;handle()
.../setup.php(14): Setup_Core::dispatchRequest()

In order for the upgrade to be successful, it was necessary for me to upgrade Tinebase by itself first, and then upgrade each application.

Having said that, I&#039;m now running through the process again (from a fresh restore of the database and I seem to be getting the error on updating tinebase every single time).

In three previous attempts (between which I restored the database back to its pre-upgrade state and ran through the upgrade again) I managed to get all applications upgraded without any visible error messages.

However, once logged in to Tine and testing the various applications I found I was not able to create any new Contacts or Tasks, due to a foreign key error related to the container_content table. Unfortunately as I don&#039;t seem to be able to get the upgrade to work again (very strange) I cannot re-produce the error again to copy down the exact details.

It would seem as though this is related to some issue throughout the upgrade as my pre-upgrade database does not contain a container_content table.

The attached image shows the version numbers of the various applications before I try to update to Joey.

**Comment by Gloirin** [comment]
**Comment posted by fvisic on 9 Nov 2012 14:21**

Did you upgrade MySQL database?
Maybe you have some table type MyISAM and some InnoDB, convert all table to InnoDB.

**Comment by Gloirin** [comment]
**Comment posted by TheSchmuck on 10 Nov 2012 01:50**

Hi fvisic,

Thankyou for looking into this so quickly. As it happens, the database I was upgrading *was* using MyISAM tables.

I have changed to using InnoDB tables by dumping and re-loading the structure and data separately. I then changed all of the CREATE TABLE statements in the structure SQL script to use InnoDB. I checked for some differences between a new Joey database and my old database and also changed some of the AUTO INCREMENT values in my structure SQL to match the newer database.

Changing to InnoDB has allowed the upgrade to get past the error I mention in the original ticket, but now I find the following error (a couple of upgrade passes later on):

SQLSTATE[HY000]: General error: 1025 Error on rename of &#039;./tine20/tine20_acsync_synckey&#039; to &#039;./tine20/#sql2-400b-16f6&#039; (errno: 152)

.../library/Zend/Db/Statement.php(284): Zend_Db_Statement_Pdo-&gt;_execute()
.../library/Zend/Db/Adapter/Abstract.php(468): Zend_Db_Statement-&gt;execute()
.../library/Zend/Db/Adapter/Pdo/Abstract.php(238): Zend_Db_Adapter_Abstract-&gt;query()
.../Setup/Backend/Abstract.php(207): Zend_Db_Adapter_Pdo_Abstract-&gt;query()
.../Setup/Backend/Abstract.php(370): Setup_Backend_Abstract-&gt;execQueryVoid()
.../ActiveSync/Setup/Update/Release5.php(32): Setup_Backend_Abstract-&gt;dropForeignKey()
.../Setup/Controller.php(435): ActiveSync_Setup_Update_Release5-&gt;update_0()
.../Setup/Controller.php(334): Setup_Controller-&gt;updateApplication()
.../Setup/Frontend/Json.php(118): Setup_Controller-&gt;updateApplications()
[internal function]: Setup_Frontend_Json-&gt;updateApplications()
.../library/Zend/Server/Abstract.php(232): call_user_func_array()
.../Zend/Json/Server.php(558): Zend_Server_Abstract-&gt;_dispatch()
.../Zend/Json/Server.php(197): Zend_Json_Server-&gt;_handle()
.../Setup/Server/Json.php(77): Zend_Json_Server-&gt;handle()
.../Setup/Core.php(89): Setup_Server_Json-&gt;handle()
.../setup.php(14): Setup_Core::dispatchRequest()

My little bit of searching suggests this error may be due to a script trying to reference a constraint that can&#039;t be found. Given I have converted a dump of a MyISAM database, which doesn&#039;t have any constraints - could my original database be missing some constraints that the upgrade script expects to be there?

Were all earlier versions of TINE configured to create MyISAM tables? Or is my situation somehow unique?

**Comment by Gloirin** [comment]
**Comment posted by TheSchmuck on 10 Nov 2012 02:09**

A little more info. I started the upgrade process over again, and went through application by application. The only failure was the above mentioned error, and it related specifically to ActiveSync. I wasn&#039;t using ActiveSync previously (even though it was enabled), so I simply uninstalled it and re-installed it to get around the upgrade problem.

However, I notice when I dump the structure of my upgraded database and compare it with a dump of a database that was created from scratch by the latest (Joey) release, the upgraded database is missing lots of constraints.

I gather these are important since they&#039;re all set to perform cascading operations.

I think I can get around this now by dumping my data only from the upgraded DB and then re-creating the database by importing the structure from the &quot;factory default joey&quot; DB and restoring my data from the separate data-only dump. 

I&#039;ll give that a go - but I thought I would mention the missing constraints in case this is an issue that will affect other upgrading users.

For your information, I have uploaded both &quot;structure dumps&quot; for you to diff. The constraints, and an auto_increment value seem to be the only differences (I haven&#039;t looked in thorough detail).

**Comment by Gloirin** [comment]
**Comment posted by TheSchmuck on 10 Nov 2012 05:41**

There seem to be even more issues with this upgrade. I&#039;ve tried the approach I mentioned earlier (use the table and constraint structure from a fresh database along with the table data from my upgrade in order that all of the necessary constraints will be in place).

Unfortunately it seems various strange things begin to happen after such an upgrade.

One is that if I uninstall and then attempt to re-install the CRM application I get the error message &quot;Sorry, your request could not be completed because the required data could not be found. In most cases this means that someone already deleted the data. Please refresh your current view&quot;.

There is no error in the PHP error log (even though error logging is switched on).

In spite of the error message, if I refresh the view, it *appears* that CRM is installed.

Then when I go to the login page for Tine (index.php) I get the dreaded &quot;The server is currently unable to handle the request due to a temporary overloading, maintenance or misconfiguration of the server. Please try again or contact your administrator.&quot; message (and no login form).

I seem to have better luck just running with the upgraded database - although I still have the issue uninstalling / installing CRM (except that I don&#039;t get the &#039;temporary overloading&#039; error locking me out of the whole system). Running the system without constraints with cascading updates worries me though.

Are you aware of many users who have successfully been upgrading their databases along the way?

**Comment by Gloirin** [comment]
**Comment posted by TheSchmuck on 10 Nov 2012 07:53**

One last update before I sit back and wait on some suggestions :) ...

This time I took a different approach and did the upgrade from the original structure and data - and *then* ran a script I prepared to put in the missing foreign key constraints.

In the process of running that script I discovered there was some data in my database that violated a couple of the foreign key constraints (some orphaned records in tine20_container_acl and tine20_preferences).

I ran some queries to delete those orphaned records, added the foreign keys and Tine *seems* to be running ok with my upgraded data. I still got the error &quot;Sorry your request could not be completed because the required data could not be found...&quot; when installing CRM, but it *seems* (after a very quick play with the CRM module) that it&#039;s working ok.

I&#039;m not sure if my database might be missing anything (there are no favourites in the CRM for instance - but perhaps it doesn&#039;t normally come with any?)

I&#039;ve attached the SQL script I used to add foreign key constraints and delete orphaned records FYI.

**Comment by Gloirin** [comment]
**Comment posted by fvisic on 10 Nov 2012 18:15**

Maybe this isn&#039;t the best solution, but you can disable all constraints with this sql command:  SET FOREIGN_KEY_CHECKS=0;

**Comment by Gloirin** [comment]
**Comment posted by TheSchmuck on 11 Nov 2012 00:40**

Hi there,

I can see why I might want to disable foreign key checks while importing data - but are you suggesting I leave the foreign key checks disabled permanently? Is that not a dangerous thing to do?

Given that the only data I am really interested in keeping are my timetracker data (timeaccounts + timesheets) and my shared tags, is there any fairly straightforward process I could use to just dump / load that data? (is it pretty self-contained in a few tables?) Then I could just start with a new Joey database and bring in the timesheets and tags.

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 19 Nov 2012 15:38**

hi TheSchmuck,

i think, fvisic suggested to turn of the checks only during the update.

and even if you run tine with disabled checks, in most of the cases it should handle foreign records correctly (without the help of db cascades, etc). the fk checks only help to ensure the integrity of the data in case of an error that is not handled.

i&#039;m not sure what we should do about this as imho yours might be a special case. and the mixing of innodb and myisam was not forced by tine.

to keep your timetracker data you would need to dump the timeaccounts, timesheets and container tables (each timeaccount has its own container for acl management). you would only need the containers assigned to the timetracker application.

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 8 Jan 2013 12:41**

this should already be fixed in the upcoming service release 2012.10.3. if you still have problems with the update, please reopen the bugreport.
